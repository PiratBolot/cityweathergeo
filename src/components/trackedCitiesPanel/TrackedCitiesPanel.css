.tracked_city {
    margin-top: 12px;
}

.tracked_city_panel {
    background-color: white;
    margin: 10px 5px 5px 5px;
    border-radius: 5px;
    box-shadow: 0 0 5px 0 #202020;
    padding: 10px;
    text-align: start;
    color: #435257;
    font-weight: bold;
    font-size: 24px;

    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
}

.add_form {
    display: flex;
    flex-direction: row;
    align-items: stretch;
}

.add_field {
    padding: 5px 10px;
    margin-right: 10px;
    background-color: #8199A3;
    border: none;
    color: white;
}

.add_field::placeholder {
    color: blanchedalmond;
}

.add_button {
    border-radius: 100%;
    height: 40px;
    width: 40px;
    background-color: #8199A3;
    font-size: medium;
    color: white;
    padding: 10px;
    border: none;
    font-weight: bold;
}

.tracked_city_list {
    display: flex;
    flex-direction: row;
    width: 100%;
    flex-wrap: wrap;
}