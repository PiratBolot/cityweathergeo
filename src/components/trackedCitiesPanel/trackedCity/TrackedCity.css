.tracked_city {
    color: #738287;
    flex-grow: 1;
    background-color: white;
    margin: 5px;
    border-radius: 5px;
    padding: 10px 0px;
    box-shadow: 0 0 5px 0 #202020;
}
