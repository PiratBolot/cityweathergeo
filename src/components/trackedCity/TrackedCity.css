.tracked_city {
    color: #738287;
    flex-grow: 1;
    background-color: white;
    margin: 5px;
    border-radius: 5px;
    padding: 10px 0;
    box-shadow: 0 0 5px 0 #202020;
}

.wf_header {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
    padding: 0 10px;
}

.city_name {
    font-size: 24px;
    font-weight: bold;
    padding: 0 5px;
}

.temperature_c {
    color: #435257;
    font-size: 40px;
    padding: 0 5px;
    line-height: 1;
}

.weather_icon {
    max-width: 40px;
    max-height: 40px;
    filter: drop-shadow(2px 4px 6px black);
}

.remove_city_button {
    display: inline-block;
    background-color: #435257;
    color: white;
    margin: 10px;
    padding: 10px;
    border: none;
    border-radius: 50%;
    box-shadow: 0 3px 20px rgba(0,0,0,.25);
}
