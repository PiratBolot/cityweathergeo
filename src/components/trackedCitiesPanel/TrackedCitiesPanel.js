import React from 'react';
import './TrackedCitiesPanel.css';

import {addTrackedCity, setErrorState, resetError} from "../../actions/Actions";
import {getWeatherByCityName} from "../../WeatherApi";
import TrackedCity from "../trackedCity/TrackedCity";
import {connect} from "react-redux";


class TrackedCitiesPanel extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            city: '',
        };

        this.addCityHandler = this.addCityHandler.bind(this);
        this.handleChange = this.handleChange.bind(this);
    }

    handleChange = (e) => {
        this.setState({city: e.target.value});
    };

    addCityHandler = async (e) => {
        e.preventDefault();
        await this.addTrackedCity(this.state.city);
        this.setState({city: ''});
    };

    addTrackedCity = async (city) => {
        if (city === "") {
            this.props.failureAddCity("Пустой запрос");
            return false;
        }
        if (!this.props.addFavoriteCity(city)) {
            this.props.failureAddCity("Город " + city + " уже отслеживается");
        }
    };

    render = () => (
        <div className="tracked_city">
            <div className="tracked_city_panel">
                <div className="panel_text">Избранное</div>
                <form className="add_form" onSubmit={this.addCityHandler}>
                    <input type="text" value={this.state.city} className="add_field"
                           onChange={this.handleChange} placeholder="Добавить новый город"/>
                    <input type="submit" className="add_button" value="+"/>
                </form>
            </div>
            <div className="tracked_city_list">
                {
                    this.props.trackedCities &&
                    this.props.trackedCities.map(
                        (e) => <TrackedCity key={e} city={e}/>
                    )
                }
            </div>
        </div>
    );
}

const mapStateToProps = (state) => ({
    trackedCities: state.trackedCities,
});

const mapDispatchToProps = (dispatch) => ({
    addFavoriteCity: (cityName) => dispatch(addTrackedCity(cityName)),
    failureAddCity: (msg) => dispatch(setErrorState(msg))
});

export default connect(mapStateToProps, mapDispatchToProps)(TrackedCitiesPanel);