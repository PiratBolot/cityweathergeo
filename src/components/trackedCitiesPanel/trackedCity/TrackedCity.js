import React from "react";
import './TrackedCity.css';

import {getWeatherByCityName, parseWeatherResponse} from '../../../WeatherApi'
import WeatherProps from "../../weatherProps/WeatherProps";
import {connect} from "react-redux";
import {deleteTrackedCity, resetError, setErrorState} from "../../../actions/Actions";
import CityProps from "../../cityProps/CityProps";
import Loader from "../../preLoader/Loader";

class TrackedCity extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            city: props.city,
            loaded: false
        };
    }

    async componentDidMount() {
        let data = await getWeatherByCityName(this.props.city);
        if (data.status === "200") {
            if (this.props.trackedCities.indexOf(this.props.city) !== -1) {
                this.props.successAddCity();
            }
        } else {
            this.props.failureAddCity(data.status === "404" ? "Город " + this.props.city + " не найден" : "Ошибка API " + data.status);
            this.props.deleteCity(this.props.city);
        }
        let res = data.response;
        let parsedData = parseWeatherResponse(res);

        this.setState({loaded: !!res, data: res, parsedWeather: parsedData, city: res.name});
    }

    render = () => (
        <div className="weather_frame">
            <CityProps isCurrentCity={false} city={this.props.city} data={this.state.data} />
            {
                this.state.loaded ?
                    <WeatherProps parsedWeather={this.state.parsedWeather} />
                    :
                    <Loader />
            }
        </div>
    )
}

const mapStateToProps = (state) => ({
    trackedCities: state.trackedCities,
});

const mapDispatchToProps = (dispatch) => ({
    deleteCity: (city) => dispatch(deleteTrackedCity(city)),
    successAddCity: () => dispatch(resetError()),
    failureAddCity: (msg) => dispatch(setErrorState(msg))
});

export default connect(mapStateToProps, mapDispatchToProps)(TrackedCity);