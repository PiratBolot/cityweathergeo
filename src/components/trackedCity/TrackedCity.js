import React from "react";
import './TrackedCity.css';

import {getWeatherByCityName, parseWeatherResponse} from '../../WeatherApi'

import CircularSpinner from "../preLoader/CircularSpinner";
import {connect} from "react-redux";
import {deleteTrackedCity} from "../../actions/Actions";

class TrackedCity extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            city: props.city,
            loaded: false
        };

        this.removeTrackedCity = this.removeTrackedCity.bind(this);
    }

    async componentDidMount() {
        let data = await getWeatherByCityName(this.props.city);
        let res = data.response;
        let parsedData = parseWeatherResponse(res);

        /* Uncomment if you want to see the preLoader explicitly
        var answer = prompt('question', 'defaultAnswer');
        */

        this.setState({loaded: !!res, data: res, parsedWeather: parsedData, city: res.name});
    }

    removeTrackedCity = async (e) => {
        e.preventDefault();
        this.props.deleteFavCity(this.state.city);
    };

    render = () => (
        <div className="weather_frame">
            <div className="wf_header">
                <div className="city_name">
                    {this.state.city}
                </div>
                {
                    this.state.data ?
                        <div className="temperature_c">
                            <span>{(this.state.data.main.temp - 273.14).toFixed(1) + "°C"}</span>
                        </div> : ""
                }
                {
                    this.state.data ?
                        <img className="weather_icon" src={"http://openweathermap.org/img/wn/" +
                        this.state.data.weather[0].icon + ".png"} alt=""
                        /> : ""
                }
                <button className="remove_city_button" onClick={this.removeTrackedCity}>X</button>
            </div>
            {
                this.state.loaded ?
                    <div className="wn_body">
                        {
                            this.state.parsedWeather.map((e) =>
                                <div key={e.key} className="wn_line">
                                    <div className="info_label">{e.key}</div>
                                    <div className="info_value">{e.value}</div>
                                </div>
                            )
                        }
                    </div>
                    :
                    <div className="image_container_overlay">
                        <p className="label">Подождите, данные загружаются</p>
                        <CircularSpinner />
                    </div>
            }
        </div>
    )
}

const mapStateToProps = (state) => ({});

const mapDispatchToProps = (dispatch) => ({
    deleteFavCity: (city) => dispatch(deleteTrackedCity(city))
});

export default connect(mapStateToProps, mapDispatchToProps)(TrackedCity);