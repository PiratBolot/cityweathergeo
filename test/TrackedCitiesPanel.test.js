import React from 'react';
import {mount} from '../enzyme/enzyme';
import {applyMiddleware, createStore} from "redux";
import {Provider} from "react-redux";
import rootReducer from "../src/reducers/Reducers";
import thunk from "redux-thunk";
import toJson from "enzyme-to-json";
import TrackedCity from "../src/components/trackedCitiesPanel/trackedCity/TrackedCity";
import TrackedCitiesPanel from "../src/components/trackedCitiesPanel/TrackedCitiesPanel";

describe('Tracked Cities Panel', () => {
    const initialState = {
        trackedCities: ['Syktyvkar', 'Moscow', 'Kiev'],
        errorMessage: {isError: false, errorMsg: ""},
        geolocation: {}
    };
    let store,wrapper;
    beforeEach(()=>{
        store = createStore(rootReducer, initialState, applyMiddleware(thunk));
        wrapper = mount( <Provider store={store}><TrackedCitiesPanel /></Provider> );
    });
    it('+++capturing Snapshot of TrackedCitiesPanel', () => {
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});