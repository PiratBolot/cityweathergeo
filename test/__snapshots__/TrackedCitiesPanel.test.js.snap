// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tracked Cities Panel +++capturing Snapshot of TrackedCitiesPanel 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(TrackedCitiesPanel)>
    <TrackedCitiesPanel
      addFavoriteCity={[Function]}
      failureAddCity={[Function]}
      trackedCities={
        Array [
          "Syktyvkar",
          "Moscow",
          "Kiev",
        ]
      }
    >
      <div
        className="tracked_city"
      >
        <div
          className="tracked_city_panel"
        >
          <div
            className="panel_text"
          >
            Избранное
          </div>
          <form
            className="add_form"
            onSubmit={[Function]}
          >
            <input
              className="add_field"
              onChange={[Function]}
              placeholder="Добавить новый город"
              type="text"
              value=""
            />
            <input
              className="add_button"
              type="submit"
              value="+"
            />
          </form>
        </div>
        <div
          className="tracked_city_list"
        >
          <Connect(TrackedCity)
            city="Syktyvkar"
            key="Syktyvkar"
          >
            <TrackedCity
              city="Syktyvkar"
              deleteCity={[Function]}
              failureAddCity={[Function]}
              successAddCity={[Function]}
              trackedCities={
                Array [
                  "Syktyvkar",
                  "Moscow",
                  "Kiev",
                ]
              }
            >
              <div
                className="weather_frame"
              >
                <CityProps
                  city="Syktyvkar"
                  isCurrentCity={false}
                >
                  <div>
                    <div
                      className="weather_now"
                    >
                      <div
                        className="city_name"
                      >
                        Syktyvkar
                      </div>
                      <Connect(RemoveCityButton)
                        city="Syktyvkar"
                      >
                        <RemoveCityButton
                          city="Syktyvkar"
                          deleteCity={[Function]}
                        >
                          <button
                            className="remove_city_button"
                            onClick={[Function]}
                          >
                            X
                          </button>
                        </RemoveCityButton>
                      </Connect(RemoveCityButton)>
                    </div>
                  </div>
                </CityProps>
                <Loader>
                  <div
                    className="image_container_overlay"
                  >
                    <p
                      className="label"
                    >
                      Подождите, данные загружаются
                    </p>
                    <CircularSpinner>
                      <div
                        className="loader"
                      >
                        <svg
                          className="circular"
                          viewBox="25 25 50 50"
                        >
                          <circle
                            className="path"
                            cx="50"
                            cy="50"
                            fill="none"
                            r="20"
                            strokeMiterlimit="10"
                            strokeWidth="2"
                          />
                        </svg>
                      </div>
                    </CircularSpinner>
                  </div>
                </Loader>
              </div>
            </TrackedCity>
          </Connect(TrackedCity)>
          <Connect(TrackedCity)
            city="Moscow"
            key="Moscow"
          >
            <TrackedCity
              city="Moscow"
              deleteCity={[Function]}
              failureAddCity={[Function]}
              successAddCity={[Function]}
              trackedCities={
                Array [
                  "Syktyvkar",
                  "Moscow",
                  "Kiev",
                ]
              }
            >
              <div
                className="weather_frame"
              >
                <CityProps
                  city="Moscow"
                  isCurrentCity={false}
                >
                  <div>
                    <div
                      className="weather_now"
                    >
                      <div
                        className="city_name"
                      >
                        Moscow
                      </div>
                      <Connect(RemoveCityButton)
                        city="Moscow"
                      >
                        <RemoveCityButton
                          city="Moscow"
                          deleteCity={[Function]}
                        >
                          <button
                            className="remove_city_button"
                            onClick={[Function]}
                          >
                            X
                          </button>
                        </RemoveCityButton>
                      </Connect(RemoveCityButton)>
                    </div>
                  </div>
                </CityProps>
                <Loader>
                  <div
                    className="image_container_overlay"
                  >
                    <p
                      className="label"
                    >
                      Подождите, данные загружаются
                    </p>
                    <CircularSpinner>
                      <div
                        className="loader"
                      >
                        <svg
                          className="circular"
                          viewBox="25 25 50 50"
                        >
                          <circle
                            className="path"
                            cx="50"
                            cy="50"
                            fill="none"
                            r="20"
                            strokeMiterlimit="10"
                            strokeWidth="2"
                          />
                        </svg>
                      </div>
                    </CircularSpinner>
                  </div>
                </Loader>
              </div>
            </TrackedCity>
          </Connect(TrackedCity)>
          <Connect(TrackedCity)
            city="Kiev"
            key="Kiev"
          >
            <TrackedCity
              city="Kiev"
              deleteCity={[Function]}
              failureAddCity={[Function]}
              successAddCity={[Function]}
              trackedCities={
                Array [
                  "Syktyvkar",
                  "Moscow",
                  "Kiev",
                ]
              }
            >
              <div
                className="weather_frame"
              >
                <CityProps
                  city="Kiev"
                  isCurrentCity={false}
                >
                  <div>
                    <div
                      className="weather_now"
                    >
                      <div
                        className="city_name"
                      >
                        Kiev
                      </div>
                      <Connect(RemoveCityButton)
                        city="Kiev"
                      >
                        <RemoveCityButton
                          city="Kiev"
                          deleteCity={[Function]}
                        >
                          <button
                            className="remove_city_button"
                            onClick={[Function]}
                          >
                            X
                          </button>
                        </RemoveCityButton>
                      </Connect(RemoveCityButton)>
                    </div>
                  </div>
                </CityProps>
                <Loader>
                  <div
                    className="image_container_overlay"
                  >
                    <p
                      className="label"
                    >
                      Подождите, данные загружаются
                    </p>
                    <CircularSpinner>
                      <div
                        className="loader"
                      >
                        <svg
                          className="circular"
                          viewBox="25 25 50 50"
                        >
                          <circle
                            className="path"
                            cx="50"
                            cy="50"
                            fill="none"
                            r="20"
                            strokeMiterlimit="10"
                            strokeWidth="2"
                          />
                        </svg>
                      </div>
                    </CircularSpinner>
                  </div>
                </Loader>
              </div>
            </TrackedCity>
          </Connect(TrackedCity)>
        </div>
      </div>
    </TrackedCitiesPanel>
  </Connect(TrackedCitiesPanel)>
</Provider>
`;
